From 94dd6b10ad9be59faeb4f21f58d7f37da7e9cf56 Mon Sep 17 00:00:00 2001
From: Elizabeth Figura <zfigura@codeweavers.com>
Date: Sat, 1 Nov 2025 18:59:23 -0500
Subject: [PATCH 35/48] server: Retain the ? suffix when renaming or linking
 reparse points.

(cherry picked from commit 6a533f21e7d26210b8d23262c910092a9eecd708)

CW-Bug-Id: #26492
---
 server/fd.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/server/fd.c b/server/fd.c
index 9043ff7eda6..9fd0f9c4e6a 100644
--- a/server/fd.c
+++ b/server/fd.c
@@ -2880,8 +2880,10 @@ static void set_fd_name( struct fd *fd, struct fd *root, const char *nameptr, da
         set_error( STATUS_OBJECT_PATH_SYNTAX_BAD );
         return;
     }
-    if (!(name = mem_alloc( len + 1 ))) return;
+    if (!(name = mem_alloc( len + 2 ))) return;
     memcpy( name, nameptr, len );
+    if (fd->unix_name[strlen( fd->unix_name ) - 1] == '?' && name[len - 1] != '?')
+        name[len++] = '?';
     name[len] = 0;
 
     if (root)
-- 
2.53.0

